lib/maths/CBoostedTreeImpl.cc

```diff
@@ -665,19 +662,21 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
     //  4. Update predictions and loss derivatives
 
     double eta{m_Eta};
-    double oneMinusBias{eta};
+    std::size_t nextTreeCountToRefreshSplits{
+        forest.size() + static_cast<std::size_t>(std::max(0.5 / eta, 1.0))};
 
-    TDoubleVecVec candidateSplits(this->candidateSplits(frame, trainingRowMask));
+    auto downsampledRowMask = this->downsample(trainingRowMask);
```

Comment:
Did you try to use your implementation of stratified sampling for regression/classification here? I would expect, it should lead to better results. My intuiting is that using stratified sampling instead of simple sampling would better preserver the data distribution properties.

Reply:
This is something I haven't explored yet. There is some prior art for stochastic gradient boosting, i.e. Friedman, which uses pure random sampling. Also, this approach can be implemented directly on the mask and doesn't require visiting the data, which is attractive from a run time perspective. As discussed offline, I'm going to raise an issue to explore knock on changes we should explore given this change and I'll mention this one.